### PR TITLE
feat(envd): memory secure token storage

### PR DESCRIPTION
**Pull request by dobrac**
> [!NOTE]
> **Medium Risk**
> Touches authentication/token validation and request parsing; mistakes could cause auth regressions or token-reset edge cases, though changes are scoped and covered by unit/integration tests.
> 
> **Overview**
> Introduces **memory-secure access token handling** in `envd` by replacing plain `*string` tokens with a new `SecureToken` backed by `memguard` (secure allocation, constant-time compare, explicit `Destroy`, and zeroing of input buffers).
> 
> Updates `/init` and auth/signing flows to avoid keeping token material in normal memory: the `/init` handler now reads the raw body to allow wiping it after JSON parsing, transfers tokens via `TakeFrom` and destroys unused tokens, and MMDS hash validation now hashes token bytes (`HashAccessTokenBytes`) with wiped copies. OpenAPI-generated clients are adjusted to use `SecureToken` types (aliased to `string` in orchestrator/integration clients), and tests are updated/added to cover secure token behavior and prevent deadlocks.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 3c4ec8a91eda0a9486a148ac930405d12d073cce. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: packages/envd/internal/api/secure_token_test.go [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+
+	var wg sync.WaitGroup
+	const numGoroutines = 100
+
+	// Concurrent reads
+	for range numGoroutines {
+		wg.Go(func() {
+			st.IsSet()
+			st.Equals("initial-token")
+		})
```
[the remark this answers is not in the document]
Comment (dobrac) [reply packages/envd/internal/api/secure_token_test.go:120]: outdated suggestion

File: packages/envd/internal/api/auth.go [hunk 2]
```diff
@@ -50,17 +50,24 @@ func (a *API) WithAuthorization(handler http.Handler) http.Handler {
...
+			tokenBytes[i] = 0
+		}
+	}()
 
 	var signature string
 	hasher := keys.NewSHA256Hashing()
 
 	if signatureExpiration == nil {
-		signature = fmt.Sprintf("%s:%s:%s:%s", path, operation, username, *a.accessToken)
+		signature = fmt.Sprintf("%s:%s:%s:%s", path, operation, username, string(tokenBytes))
```
Comment (djeebus) [inline packages/envd/internal/api/auth.go]: Not your line originally, but might be more readable to do `strings.Join([]string{path, operation, username, string(tokenBytes)}, ":")`

File: packages/envd/internal/api/secure_token.go [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+	defer s.mu.Unlock()
+
+	// Destroy old token first (zeros memory)
+	if s.buffer != nil {
+		s.buffer.Destroy()
+		s.buffer = nil
+	}
+
+	// Create new LockedBuffer from bytes (source is wiped automatically)
+	s.buffer = memguard.NewBufferFromBytes([]byte(token))
```
Comment (djeebus) [inline packages/envd/internal/api/secure_token.go]: What does "source is wiped automatically" mean? If we were to add an `if token != "" { panic("this did not work" }` after this line, would it not panic?

Comment (dobrac) [reply packages/envd/internal/api/secure_token.go]: very good point. The []byte() is cleaned, but the token remains. I'll refactor it to be already serialized from JSON to the memguarded memory

Comment (dobrac) [reply packages/envd/internal/api/secure_token.go]: reworked to parsing the JSON directly to the SecureToken and wiping the buffer after. Minimizing a bit more the trace

File: packages/envd/internal/api/auth.go (hunk 2 above)
Comment (dobrac) [reply packages/envd/internal/api/auth.go]: updated

File: packages/envd/internal/api/secure_token.go [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+// UnmarshalJSON implements json.Unmarshaler to securely parse a JSON string
+// directly into memguard, wiping the input bytes after copying.
+//
+// Access tokens are hex-encoded HMAC-SHA256 hashes (64 chars of [0-9a-f]),
+// so they never contain JSON escape sequences.
+func (s *SecureToken) UnmarshalJSON(data []byte) error {
+	// JSON strings are quoted, so minimum valid is `""` (2 bytes).
+	if len(data) < 2 || data[0] != '"' || data[len(data)-1] != '"' {
+		return errors.New("invalid secure token JSON string")
+	}
```
[the remark this answers is not in the document]
Comment (dobrac) [reply packages/envd/internal/api/secure_token.go:59]: its alright.

File: packages/envd/internal/api/init.go [hunk 5]
```diff
@@ -36,9 +37,11 @@ const (
 // validateInitAccessToken validates the access token for /init requests.
 // Token is valid if it matches the existing token OR the MMDS hash.
 // If neither exists, first-time setup is allowed.
-func (a *API) validateInitAccessToken(ctx context.Context, requestToken *string) error {
+func (a *API) validateInitAccessToken(ctx context.Context, requestToken *SecureToken) error {
+	requestTokenSet := requestToken != nil && requestToken.IsSet()
+
 	// Fast path: token matches existing
-	if a.accessToken != nil && requestToken != nil && *requestToken == *a.accessToken {
+	if a.accessToken.IsSet() && requestTokenSet && a.accessToken.EqualsSecure(requestToken) {
```
Comment (ValentaTomas) [inline packages/envd/internal/api/init.go:44]: Btw, for the equal, it looks like stdlib might have some of the functionality already https://pkg.go.dev/crypto/subtle

Comment (ValentaTomas) [inline packages/envd/internal/api/init.go:44]: Both request and access tokens are hashes inside sandbox, right?

Comment (dobrac) [reply packages/envd/internal/api/init.go:44]: the EqualsSecure is not my implementation of the crypto, but rather than concurrency safe usage of the `memguard.LockedBuffer` EqualTo method

Comment (dobrac) [reply packages/envd/internal/api/init.go:44]: no, they're raw tokens unfortunately as of now

File: packages/envd/internal/api/secure_token.go [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+	return s.buffer.EqualTo(other.buffer.Bytes())
+}
+
+// IsSet returns true if a token is stored.
+func (s *SecureToken) IsSet() bool {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+
+	return s.buffer != nil && s.buffer.IsAlive()
+}
```
Comment (djeebus) [inline packages/envd/internal/api/secure_token.go:173]: Would it clean things up to make this struct nil-pointer-receiver-safe? 

```golang
func (s *SecureToken) IsSet() bool {
	if s == nil { return false }

	// implementation details
}
```

That way we can do things like this:

```golang
// old way
if tok != nil && tok.IsSet() { ... }

// new way
if tok.IsSet() { ... }
```

Not necessary, just a thought.

File: packages/envd/internal/api/secure_token.go [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+	}
+
+	content := data[1 : len(data)-1]
+
+	// Access tokens are hex strings - reject if contains backslash
+	if bytes.ContainsRune(content, '\\') {
+		memguard.WipeBytes(data)
+
+		return errors.New("invalid secure token: unexpected escape sequence")
+	}
```
Comment (djeebus) [inline packages/envd/internal/api/secure_token.go:68]: Why do we need to prevent this situation?

File: packages/envd/internal/api/init.go (hunk 5 above)
Comment (ValentaTomas) [reply packages/envd/internal/api/init.go:44]: Will we be able to switch? That should be possible based on the envd version, right?

Comment (dobrac) [reply packages/envd/internal/api/init.go:44]: We'll be absolutely able to switch to store only the hashed token. If only sending the hash, probably as you're saying based on the envd version, should be fine.

File: packages/envd/internal/api/secure_token.go (hunk 6 above)
Comment (dobrac) [reply packages/envd/internal/api/secure_token.go:173]: let me try that, that sounds like a good idea

File: packages/envd/internal/api/secure_token.go (hunk 7 above)
Comment (dobrac) [reply packages/envd/internal/api/secure_token.go:68]: JSON escape sequences require processing: JSON strings can contain escape sequences like \", \\, \n, \uXXXX, etc. Normally you'd use strconv.Unquote to decode these, but that allocates new memory. We want to copy the token directly into memguard-protected memory without intermediate allocations.

-> The backslash check lets us safely skip JSON unescaping logic while ensuring we reject any input that would require it. It's a security optimization specific to the hex-only token format.

File: packages/envd/internal/api/secure_token.go (hunk 6 above)
Comment (dobrac) [reply packages/envd/internal/api/secure_token.go:173]: updated
